Test solid mechanics with patch tests

btalamini opened the pull request:
This PR was spurred by #754. The main change is to add more rigorous tests of the solid mechanics module. These are patch tests, which ensure that the numerical approximation of the PDEs are consistent. The structure should also allow generalization to manufactured solutions, which would allow us to increase our meaningful test coverage and to verify high-order convergence.

I can see that these tests could benefit from some refactoring to reduce boilerplate, to increase flexibility of options (such as geometric nonlinearity on/off, material type), and to make testing over the combinations of these options more automatic. I'm looking for help here from the code review.

Once we settle on the framework, it should be reused to check other features:

1. Dynamics
2. Potentially, other flavors of statics options: linear kinematics, parameterized models, and materials with state
3. The heat conduction module

We may want to leave issue #754 open until we're satisfied with the coverage, possibly amending it with a checklist.

These tests have already caught 3 significant bugs:

1. The residual calculation contracted the wrong index of the stress with the shape function derivatives. A stopgap solution is put in this PR that makes the residual correct, and issue #773 was opened to track a long-term fix.
2. Several of the material models were templated on spatial dimension, but the 2D versions gave nonsensical results. The stress returned was neither plane strain nor plane stress. I removed the template parameter on the material and moved it to the constitutive output call, with the 2D version giving the plane strain response. The comments have been updated to reflect this.
3. The geometric nonlinearity option was causing the reference configuration to sometimes be in an incorrect state. This led to issue #760 (that has since been closed, thanks @jamiebramwell).

btalamini committed:
WIP Make patch test fixture

btalamini committed:
Merge branch 'develop' into testing/talamini/patch_tests

btalamini committed:
Refactor solution out of test to make it a bit reusable. Still doesn't work

btalamini committed:
Merge branch 'develop' into testing/talamini/patch_tests

btalamini committed:
Switch linear solver to be direct to remove tolerance uncertainty

btalamini committed:
Try a linear elastic patch test to simplify and isolate problems

samuelpmishLLNL committed:
get Patch2Dlinear test converging (failing)

samuelpmishLLNL committed:
give descriptive names to the boundary attributes

btalamini committed:
Switch linear solver in linear elastic problem to direct solver

btalamini committed:
Merge branch 'develop' into testing/talamini/patch_tests

btalamini committed:
Get build working again after merging

btalamini committed:
Remove dimensionally dependent material model implementations

Several material models were template parameterized on the spatial
dimensionality (2 or 3), with the implementation changing based on the
dimensionality. The 2D versions were incorrect, being neither plane
strain nor plane stress.

Now: spatial dimension templating of class removed. Parameterize
operator() on spatial dimension, such that in 2D plane strain is
effected (for the in-plane components of the stress).

btalamini committed:
Fix broken bou dary attribute in test

btalamini committed:
Make linear elastic patch test pass (finally)

A lot of numbers are hard coded that should be put in analytic
form for generality (and to make the test understandable).

btalamini committed:
WIP 3D patch, may be error in solver algorithm for inhomogeneous Dirichlet case

jamiebramwell committed:
bug: Cast the stiffness matrix to DSuperLU form when direct solvers are requested

samuelpmishLLNL committed:
fix affine function signature in tests

btalamini committed:
Get nonlinear patch test working

btalamini committed:
Merge branch 'develop' into testing/talamini/patch_tests

btalamini committed:
Make work with merge, now that ref config doesn't change

btalamini committed:
Merge branch 'bugfix/bramwell/direct_solver' into testing/talamini/patch_tests

btalamini committed:
Revert broken direct solver changes, recent merge fixes this

btalamini committed:
Remove debug output and test over several element orders

btalamini committed:
Remove old test cases that are covered by better patch tests

There are still some old tests remaining that the conditions
in the patch test don't cover yet, such as parameterized materials
and dynamics.

btalamini committed:
Remove linear elastic-only patch test

Code duplication. We can parameterize the patch test on material
if we want to do this.

btalamini committed:
Factor out loading from patch test

Factor out common framework of patch test in a function, with a
functor argument that sets the boundary conditions. Allow test writer to
reuse the same framework with different test conditions.

btalamini committed:
Refactor exact solution so it can be used to generate traction load cases

btalamini committed:
Add traction patch test

Currently fails.

btalamini committed:
Found bug in serac using patch tests, but haven't repaired it yet

Seems to be a transposition error when applying tractions.

btalamini committed:
Push temporary fix to incorrectly transposed stress

btalamini committed:
Use unstructured meshes for more rigor

Meshes have quad/hex elements with non-affine mappings from reference
element. This also tests the isoparameteric mapping in the patch test.

btalamini committed:
Refactor to reuse test for any exact solution

btalamini committed:
Reduce boilerplate for different boundary conditions

btalamini committed:
Write docstrings and tidy

btalamini committed:
Rename some symbols for clarity

btalamini committed:
Make traction test harder in 3D so that iteration is required

btalamini committed:
Rename for clarity

btalamini committed:
Add a utility function to avoid ugly call to low-level mfem method

Add missing functionality to FiniteElementVector: computing L2
error norms.

btalamini committed:
Move utility funciton from test file to better central location

btalamini committed:
Move the patch tests to own test file to make migration easier

The old norm based tets, which will eventually be removed, were
interleaved with the new patch tests.

* The PR will be easier to review and understand as an added file,
  rather than a massively edited old file.

* The existing test file names are vague, in my opinion, making it
  hard to guess where tests of a particular thing are going to be.
  This attempts to improve the naming convention.

btalamini committed:
Editing for clarity

btalamini committed:
Removed unneeded headers

btalamini committed:
Merge branch 'develop' into testing/talamini/patch_tests

btalamini committed:
Fix spacing format problems

samuelpmishLLNL commented on src/serac/physics/state/finite_element_vector.hpp:
```diff
@@ -370,4 +370,13 @@ double min(const FiniteElementVector& fe_vector);
+/**
+ * @brief Find the L2 norm of the error of a finite element vector with respect to an exact solution
+ *
+ * @param state The state variable to compute a norm of
```
it looks like this description of `state` was copied from `norm()` and needs to be updated. I never know what to put down for doxygen descriptions of arguments like this, almost everything ends up sounding tautological.

btalamini replied:
Thanks. I just updated it. I agree, but tautological is better than wrong. In a way, it's a good sign, because it means the function is probably self-documenting.

samuelpmishLLNL commented on src/serac/physics/tests/test_solid_functional_statics.cpp:
```diff
@@ -0,0 +1,290 @@
+    A(0, 0) = 0.110791568544027; A(0, 1) = 0.230421268325901;
+    A(1, 0) = 0.198344644470483; A(1, 1) = 0.060514559793513;
+    if constexpr (dim == 3) {
+      A(0, 2) = 0.15167673653354;
+      A(1, 2) = 0.084137393813728;
+      A(2, 0) = 0.011544253485023; A(2, 1) = 0.060942846497753; A(2, 2) = 0.186383473579596;
```
I think I see the point of this formatting, to make the code graphically represent the matrix terms that are affected. It could also be possible to do something like:
```suggestion
                                                                A(0, 2) = 0.15167673653354;
                                                                A(1, 2) = 0.084137393813728;
      A(2, 0) = 0.011544253485023; A(2, 1) = 0.060942846497753; A(2, 2) = 0.186383473579596;
```

Also, if the values of these entries is not important, then it's possible to make this initialization a little more succinct (using a different arbitrary choice of values)
```cpp
for (int i = 0; i < dim; i++) {
  for(int j = 0; j < dim; j++) {
    A(i,j) = 1.0 / (i + 2 * j);
  }
  b[i] = sin(i);
}
```
would have a similar effect

btalamini replied:
This is a good suggestion. In the long run, it would be even better to make the essential boundary condition calls use `tensor`, which would make this code more compact and readable. It would also increase the uniformity of the interface with the other loading functions.

samuelpmishLLNL commented on src/serac/physics/tests/test_solid_functional_statics.cpp:
```diff
@@ -0,0 +1,290 @@
+   */
+  void operator()(const mfem::Vector& X, mfem::Vector& u) const
+  {
+    u = 0.0;
+    A.Mult(X, u);
```
I think `Operator::Mult()` zeroes out the output-vector for us (?), which would make line 62 is unnecessary

btalamini replied:
Thanks. You're right. Fixed.

samuelpmishLLNL commented on src/serac/physics/tests/test_solid_functional_statics.cpp:
```diff
@@ -0,0 +1,290 @@
+ * @param b Kind of boundary conditions to apply in the problem
+ * @return std::set<int> Boundary attributes for the essential boundary condition
+ */
+template <int dim>
+std::set<int> essentialBoundaryAttributes(PatchBoundaryCondition bc)
+{
+  if constexpr (dim == 2) {
```
It's okay to pass `dim` as a template parameter, but this function doesn't appear to really need compile-time information of the value of `dim`. So, it could be passed just as well as a conventional argument (this would require changing the `if constexpr`s into regular `if`s). Either implementation is acceptable as far as I'm concerned.

btalamini replied:
Good point. I'll leave it for now, just so the `static_assert` on the spatial dimension in the test body doesn't break. Not sure that is the right place to test this condition, but that's probably a subject for another PR.

samuelpmishLLNL commented on data/meshes/patch3D.mesh:
```diff
@@ -0,0 +1,71 @@
+1.5 1.5 1.5
+0.5 1.5 1.6
+
+#0 0 0
+#1 0 0
+#1 1 0
+#0 1 0
+#0 0 1
+#1 0 1
+#1 1 1
+#0 1 1
+#0.25 0.333333333333333 0.25
+#0.62 0.25 0.25
+#0.75 0.75 0.25
+#0.25 0.75 0.25
+#0.25 0.25 0.75
+#0.75 0.25 0.75
+#0.75 0.75 0.75
+#0.25 0.75 0.8
```
Can we delete these extra coordinates?

btalamini replied:
Oops, that was careless. Deleted them.

samuelpmishLLNL commented:
For reference, a visualization of the 3D patch test mesh is the red version below:
![patch_test_mesh](https://user-images.githubusercontent.com/61714427/186518110-09930944-a049-416d-a1c5-4a4c956f4516.gif)

In contrast to the blue version which has a lot of symmetries, the red version has the interior nodes perturbed to break the symmetries and ensure that the element geometries are not axis-aligned.

btalamini committed:
Remove commented junk from mesh file

btalamini committed:
Format cleanup

Co-authored-by: samuelpmishLLNL <61714427+samuelpmishLLNL@users.noreply.github.com>

btalamini committed:
Remove redundant code

MFEM already zeros out output vectors in Mult() calls

jamiebramwell commented on src/serac/physics/tests/test_solid_functional_statics.cpp:
```diff
@@ -0,0 +1,289 @@
+  axom::sidre::DataStore datastore;
+  serac::StateManager::initialize(datastore, "solid_functional_static_solve");
+
+  // BT: shouldn't this assertion be in the physics module?
```
Agreed. I'll add an issue.

jamiebramwell reviewed (approved):
This is fantastic (and well past due)! Thank you for doing this!

btalamini committed:
Fix doxygen comment

btalamini committed:
Fix clang warning

btalamini committed:
Amend an incomplete change in previous commit

btalamini committed:
make style

btalamini committed:
Fix case of filename

btalamini committed:
Make patch tests serial to avoid SuperLU error on very small system

btalamini committed:
Remove boundary condition test that is now covered by a more rigorous test

btalamini committed:
Re-baseline placeholder tests for now

* The diffs are expected due to the bug fixes prompted by the
  patch testing. Leave these tests for now as placeholders since
  they cover features not tested by real patch tests yet.
  These tests will be supplanted by additional patch tests.
* Remove one test that is no longer needed. (traction test). It is
  alreay tested more rigorously by path tests.

btalamini committed:
make style

btalamini committed:
Merge branch 'develop' into testing/talamini/patch_tests
